src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs

```diff
@@ -113,11 +114,55 @@ internal static bool IsAllNullBytes(Span<byte> buffer)
             return true;
         }
 
-        // Returns a DateTimeOffset instance representing the number of seconds that have passed since the Unix Epoch.
-        internal static DateTimeOffset GetDateTimeFromSecondsSinceEpoch(double secondsSinceUnixEpoch)
+        // Converts the specified number of seconds that have passed since the Unix Epoch to a DateTimeOffset.
+        internal static DateTimeOffset GetDateTimeOffsetFromSecondsSinceEpoch(double secondsSinceUnixEpoch) =>
+            new DateTimeOffset((long)(secondsSinceUnixEpoch * TimeSpan.TicksPerSecond) + DateTime.UnixEpoch.Ticks, TimeSpan.Zero);
+
+        // Converts the specified DateTimeOffset to the number of seconds that have passed since the Unix Epoch.
+        internal static double GetSecondsSinceEpochFromDateTimeOffset(DateTimeOffset dateTimeOffset) =>
+            ((double)(dateTimeOffset.UtcDateTime - DateTime.UnixEpoch).Ticks) / TimeSpan.TicksPerSecond;
+
+        // If the specified fieldName is found in the provided dictionary and it is a valid double number, returns true and sets the value in 'dateTimeOffset'.
+        internal static bool TryGetDateTimeOffsetFromTimestampString(IReadOnlyDictionary<string, string> dict, string fieldName, out DateTimeOffset dateTimeOffset)
+        {
+            dateTimeOffset = default;
+            if (dict.TryGetValue(fieldName, out string? value) &&
+                double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out double secondsSinceEpoch))
```

Comment:
I'm confused - who uses this as a `double`? Isn't this normally an integer value? (`int` or `long`)?

Reply:
I had the same question, and found out that GNU TAR writes the timestamp as `UNIX_TIMESTAMP.SUB_MAYBE_NANO_SECONDS`, meaning it's not _really_ a `double` it's two concatenated `ulong` values (and... some answer on 32-bit systems... `uint.ulong`?.

At the big end, DateTimeOffset.MaxValue is 253402300799, which us less than 2^38, so loss of precision past 2^53 isn't a factor.  On the small end, our DateTimeOffset's unit of precision is 100ns.  Since we're currently using 30 significant bits left of the decimal that leaves 23 bits for the right, which I think means we're possibly losing about one tenth of a tick, and it'll be greater loss in 2038.

So... OK, maybe it should use a custom ulong.ulong parsing.

Reply:
It is stored as a double in the PAX extended attributes, dot included.

In the old field in the header, it is stored as integer.

Reply:
Is there documentation anywhere on this? Is it truly a `double`? Or is it like above - `ulong`.`ulong`?

Reply:
I am not sure the second number is nanoseconds, I think it's milliseconds, the next unit after seconds.

I don't even know how to add _nanoseconds_ to a `DateTimeOffset`, or if it is even possible. I was thinking of adding `TimeSpan.FromNanoSeconds` but we don't have such method, only Milli and Micro.

Reply:
This might take me longer to address, and since it's not related to the original issue, I'd like to address it separately: https://github.com/dotnet/runtime/issues/70801

Reply:
@tarekgh helped me figure it out: The original parsing is correct, we are dealing with an actual `double`. The decimal part is _fractions of seconds_.

Here is a stackoverflow thread asking about such timestamp: https://stackoverflow.com/questions/21696323/epoch-time-with-dot-1391759952-7056 . Notice that the decimal part is larger than 1000, which means it cannot be milliseconds. And the answer below is "fractions of seconds".

Also, the example in the FreeBSD tar manual is: `25 ctime=1084839148.1212\n` which again, means it is not milliseconds.